core/server/api_container/server/startosis_engine/kurtosis_types/service_config/image_build_spec.go

```diff
@@ -45,13 +46,17 @@ func NewImageBuildSpecType() *kurtosis_type_constructor.KurtosisTypeConstructor
 						return builtin_argument.NonEmptyString(value, BuildContextAttr)
 					},
 				},
+				{
+					Name:              BuildFileAttr,
+					IsOptional:        true,
+					ZeroValueProvider: builtin_argument.ZeroValueProvider[starlark.String],
+					Validator:         nil,
+				},
 				{
 					Name:              TargetStageAttr,
 					IsOptional:        true,
 					ZeroValueProvider: builtin_argument.ZeroValueProvider[starlark.String],
-					Validator: func(value starlark.Value) *startosis_errors.InterpretationError {
```

Comment:
Not sure about that @mieubrisse since I'm not very familiar with the code base

Reply:
oops didn't catch this; yeah let's not change the TargetStageAttr validator for now. I _think_ you should even be able to apply the same validator to the build-file arg too

Reply:
Alright, let me provide some context.

The tests for `ImageBuildSpec` were initially designed to cover scenarios where `target_stage` is specified. Given that `target_stage` is an optional argument and can also be empty, I introduced a test to address this case. However, the test did not pass. The issue lies in the fact that the validator expects `TargetStageAttr` to be non-empty, conflicting with its optional nature. To resolve this conflict, I set the validator for `TargetStageAttr` to `nil`. I followed the same approach for the newly introduced argument, `BuildFileAttr`.

Below are the logs I encountered when configuring the `TargetStageAttr` validator to `builtin_argument.NonEmptyString`:

```sh
~/kurtosis/core/server $ go test -v ./... 
....
RUN   TestKurtosisTypeConstructorSuite/TestImageBuildSpecTest
    suite_kurtosis_type_constructor_test.go:54: 
        	Error Trace:	/Users/leovct/Documents/opensource/kurtosis/core/server/api_container/server/startosis_engine/kurtosis_starlark_framework/test_engine/suite_kurtosis_type_constructor_test.go:54
        	            				/Users/leovct/Documents/opensource/kurtosis/core/server/api_container/server/startosis_engine/kurtosis_starlark_framework/test_engine/image_build_spec_framework_test.go:30
        	Error:      	Expected nil, but got: &starlark.EvalError{Msg:"Cannot construct 'ImageBuildSpec' from the provided arguments.\n\tCaused by: The following argument(s) could not be parsed or did not pass validation: {\"target_stage\":\"Value for 'target_stage' was an empty string. This is disallowed\"}", CallStack:starlark.CallStack{starlark.CallFrame{Name:"<toplevel>", Pos:syntax.Position{file:(*string)(0xc0000febc0), Line:1, Col:24}}, starlark.CallFrame{Name:"ImageBuildSpec", Pos:syntax.Position{file:(*string)(0x2b610a0), Line:0, Col:0}}}, cause:(*startosis_errors.InterpretationError)(0xc00029cd40)}
        	Test:       	TestKurtosisTypeConstructorSuite/TestImageBuildSpecTest
        	Messages:   	Error interpreting Starlark code. Code was: 
        	            	result = ImageBuildSpec(image_name="kurtosistech/example-datastore-server", build_context_dir="./", build_file="", target_stage="")
    mock_package_content_provider.go:519: FAIL:	GetAbsoluteLocator(string,string,string,map[string]string)
        		at: [/Users/leovct/Documents/opensource/kurtosis/core/server/api_container/server/startosis_engine/startosis_packages/mock_package_content_provider.go:161 /Users/leovct/Documents/opensource/kurtosis/core/server/api_container/server/startosis_engine/kurtosis_starlark_framework/test_engine/image_build_spec_framework_test.go:21]
    mock_package_content_provider.go:519: FAIL:	GetOnDiskAbsolutePackageFilePath(string)
        		at: [/Users/leovct/Documents/opensource/kurtosis/core/server/api_container/server/startosis_engine/startosis_packages/mock_package_content_provider.go:325 /Users/leovct/Documents/opensource/kurtosis/core/server/api_container/server/startosis_engine/kurtosis_starlark_framework/test_engine/image_build_spec_framework_test.go:26]
    mock_package_content_provider.go:519: FAIL: 0 out of 2 expectation(s) were met.
        	The code you are testing needs to make 2 more call(s).
        	at: [/Users/leovct/Documents/opensource/kurtosis/core/server/api_container/server/startosis_engine/startosis_packages/mock_package_content_provider.go:519 /Users/leovct/.asdf/installs/golang/1.20.14/go/src/testing/testing.go:1150 /Users/leovct/.asdf/installs/golang/1.20.14/go/src/testing/testing.go:1328 /Users/leovct/.asdf/installs/golang/1.20.14/go/src/testing/testing.go:1570 /Users/leovct/.asdf/installs/golang/1.20.14/go/src/runtime/panic.go:522 /Users/leovct/.asdf/installs/golang/1.20.14/go/src/testing/testing.go:980 /Users/leovct/Documents/opensource/kurtosis/core/server/api_container/server/startosis_engine/kurtosis_starlark_framework/test_engine/suite_kurtosis_type_constructor_test.go:54 /Users/leovct/Documents/opensource/kurtosis/core/server/api_container/server/startosis_engine/kurtosis_starlark_framework/test_engine/image_build_spec_framework_test.go:30]
```

Reply:
Ah, thanks for the context @leovct ! Let me tag in @tedim52 who has context on target stage

Reply:
Hey @leovct ! Thanks for adding a test for target stage. 

The behavior of the validator function is that if the argument is set (optional or not), the validator function is triggered. So if `target_stage` is not set at all, thats fine because it's optional, but if it is set, then we require it not to be the empty string. This behavior should be the exact same for `build_file`. 

To get that behavior, you can add back the original validator for `target_stage` and add it to `build_file` as well. Then for the tests, add a test where `target_stage` is empty and an error occurs (tests the validation) and a test where `target_stage` is non empty (should succeed) and repeat for `build_file`. LMK if this makes sense. Thanks again for improving our testsuite : )

Reply:
That makes sense, thank you! I reverted the change on the validator functions and fixed the tests :)